### PR TITLE
Test Unbox instances and generic derived instances manually

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fixes:https://github.com/composewell/streamly/issues/2215

File: test/Streamly/Test/Data/Unbox.hs [hunk 1]
```diff
@@ -70,10 +166,21 @@ tests = do
...
 
+testsManual :: Spec
+testsManual = do
+    prop "Int type"  $ testInt (WInt 5)
+    prop "Int8 type" $ testInt8 (WInt8 127)
+    prop "Int16 type" $ testInt16 (WInt16 444)
+    prop "Int32 type" $ testInt32 (WInt32 7777)
+    prop "Float type" $ testFloat (WFloat 33.556)
+    prop "Double type" $ testDouble (WDouble 336.588856)
+    prop "Word32 type" $ testWord32 (WWord32 4294967295)
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:177]: We discussed that we will write tests only for non-primitive types. These are all primitive types, there is not much point writing tests for these, they are only testing what the `Single` test is doing, it is testing a single wrapper type.

We need to test only those instances which are not primitive i.e. those that can be derived by generic deriving using the primitive type instances.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: test/Streamly/Test/Data/Unbox.hs [hunk 2]
```diff
@@ -46,6 +65,83 @@ test v1 v2 = monadicIO $ do
     assert (x2 == v2)
     assert (x1 == v1)
 
+testInt :: IntWrap -> Property
+testInt val =
+    monadicIO $ do
+        var <- run $ newUnpinnedBytes (sizeOf (Proxy :: Proxy IntWrap))
+        run $ pokeWith var 0 val
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:72]: Do not use peekWith/pokeWith - these are to be removed, there is an issue about it.

File: test/Streamly/Test/Data/Unbox.hs [hunk 3]
```diff
@@ -37,6 +56,165 @@ data NestedSOP =
...
      | NSOP1 SumOfProducts deriving (Unbox, Show, Generic, Eq)
 
+data NonPrim2 = NonPrim2 (StablePtr NestedSOP) deriving (Unbox, Generic, Eq)
+
+testStablePtr :: Property
+testStablePtr =
+    monadicIO $ do
+        assert ( sizeOf (Proxy :: Proxy NonPrim2)
+               == genericSizeOf (Proxy :: Proxy NonPrim2)
+                )
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:66]: Test the size of StablePtr, rather than the size of a wrapper around StablePtr.

Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:66]: Use hspec expectation instead of using a Property test.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Comment by rnjtranjan** [comment]
There is still issue with genericPeekByteIndex for Sum type:
uncaught exception: ErrorCall
       **skipByte: reading beyond limit**
       CallStack (from HasCallStack):
         error, called at src/Streamly/Internal/Data/Unboxed.hs:544:29 in streamly-core-0.1.0-inplace:Streamly.Internal.Data.Unboxed

**Comment by rnjtranjan** [comment]
2) Data.Unbox, Manual tests, Bool type
       uncaught exception: ErrorCall
       skipByte: reading beyond limit
       CallStack (from HasCallStack):
         error, called at src/Streamly/Internal/Data/Unboxed.hs:544:29 in streamly-core-0.1.0-inplace:Streamly.Internal.Data.Unboxed

File: test/Streamly/Test/Data/Unbox.hs [hunk 4]
```diff
@@ -1,4 +1,11 @@
-{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE StandaloneDeriving, DeriveAnyClass #-}
+{-# LANGUAGE DataKinds #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+{-# HLINT ignore "Use newtype instead of data" #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-simplifiable-class-constraints #-}
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:8]: Can put all the OPTIONS_GHC together.

File: test/Streamly/Test/Data/Unbox.hs [hunk 5]
```diff
@@ -37,43 +59,143 @@ data NestedSOP =
...
-    ref <- run $ IORef.newIORef v1
-    x1 <- run $ IORef.readIORef ref
-    run $  IORef.writeIORef ref v2
-    x2 <- run $  IORef.readIORef ref
-    assert (x2 == v2)
-    assert (x1 == v1)
+-- Write v1 to an Unboxed ref, read it back and compare.
+-- Then do the same for v2.
+test ::(Unbox a, Eq a, Show a) => a -> a -> IO ()
+test v1 v2 = do
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Unbox.hs:65]: We are testing `v1` independent of `v2`.
Why do we have both `v1` and `v2` here?

File: test/Streamly/Test/Data/Unbox.hs [hunk 6]
```diff
@@ -37,43 +59,143 @@ data NestedSOP =
...
+    sizeOf (Proxy :: Proxy (Ratio Int))
+            `shouldBe` genericSizeOf (Proxy :: Proxy (Ratio Int))
+    let val :: (Ratio Int) = 5 :% 3
+    var <- newUnpinnedBytes (sizeOf (Proxy :: Proxy (Ratio Int)))
+    pokeByteIndex 0 var val
+    genericPeekByteIndex var 0 `shouldReturn` val
+    genericPokeByteIndex var 0 val
+    peekByteIndex 0 var `shouldReturn` val
+
+-- XXX not working for ghc 8.6.5
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:141]: Could have put this under a GHC version checking CPP macro.

File: test/Streamly/Test/Data/Unbox.hs (hunk 5 above)
Comment (harendra-kumar) [reply test/Streamly/Test/Data/Unbox.hs:65]: I guess this is to use the same IORef to read/write two different values.

Comment (adithyaov) [reply test/Streamly/Test/Data/Unbox.hs:65]: Are we testing `IORef`?

File: test/Streamly/Test/Data/Unbox.hs [hunk 7]
```diff
@@ -16,186 +16,204 @@
...
-    it "Ratio type" testRatio
-    -- XXX not working for ghc 8.6.5
-    --it "Fingerprint type" testFingerprint
-    it "Const type" testConst
-    it "Identity type" testIdentity
+--------------------------------------------------------------------------------
+-- CPP helpers
+--------------------------------------------------------------------------------
+
+#define TEST_SERIALIZATION(def, val) it def $ testSerialization (val)
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Unbox.hs:153]: Can we use a Haskell function instead?

**Review by harendra-kumar — APPROVED** [review]

File: test/Streamly/Test/Data/Unbox.hs (hunk 7 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Unbox.hs:153]: We will

File: test/Streamly/Test/Data/Unbox.hs (hunk 6 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Unbox.hs]: See https://github.com/composewell/streamly/issues/2295

**Pull request merged**